io/io/inc/TFileMergeServer.h

```diff
@@ -0,0 +1,85 @@
+#ifndef ROOT_TFileMergeServer
+#define ROOT_TFileMergeServer
+
+#include "TBits.h"
+#include "TFile.h"
+#include "TFileCacheWrite.h"
+#include "TFileMerger.h"
+#include "TTimeStamp.h"
+
+#include <condition_variable>
+#include <exception>
+#include <stdexcept>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace ROOT {
+namespace Experimental {
+
+class TFileMergeClient;
+
+class TFileMergeServer {
+   public:
+      TFileMergeServer() = delete;
+      TFileMergeServer(const TFileMergeServer&) = delete;
+
+      TFileMergeServer(const std::string& name) : fName(name)
+      {
+         fMergingThread.reset(new std::thread([this]() { this->Listen(); }));
+      }
+
+      ~TFileMergeServer()
+      {
+         try {
+            for (auto client : fClients)
+               if(!client.expired())
+                  throw std::runtime_error("TFileMergeClients must be destroyed before the server");
+         } catch(std::runtime_error& e) {
+            std::terminate();
+         }
+
+         if (finish == false)
+            Finish();
+         fMergingThread->join();
+      }
+
+      std::string GetOutputName() const { return fName; }
+
+      std::shared_ptr<TFileMergeClient> GetClient();
+
+      void Push(TBufferFile* buffer)
+      {
+         {
+            std::lock_guard<std::mutex> lock(fQueueMutex);
```

Comment:
This is a matter of preference, I'd rather leave it as is. I prefer to use the scope to show the duration of the lock. It's also better to avoid bugs from explicitly calling lock()/unlock().